sample-service/src/main/java/uk/ac/ox/ndph/mts/sample_service/security/authorisation/AuthorisationService.java

```diff
@@ -109,14 +140,29 @@ private boolean hasRequiredPermissionInRole(RoleDTO role, String requiredPermiss
                 .anyMatch(permission -> permission.getId().equals(requiredPermission));
     }
 
-    /**
-     * Validate user site trees allowed the entity site
-     * @param userRole - participant role which includes the site property
-     * @param entitySite - the requested entity site
-     * @return true - currently it is a stub/infra until the implementation will be added
-     */
-    @SuppressWarnings("squid:S1172") //suppress unused parameter
-    private boolean isSiteAuthorised(String entitySite, String userRole) {
-        return true;
+    private boolean isAuthorisedToAllSites(List<SiteDTO> sites,
+                                          List<RoleAssignmentDTO> roleAssignments,
+                                          List<String> entitiesSiteIds) {
+
+        Map<String, ArrayList<SiteDTO>> tree = siteTreeUtil.getSiteSubTrees(sites);
+
+        var hasAnUnauthorisedSite = entitiesSiteIds.stream().anyMatch(siteId ->
```

Comment:
This will grow as O(n*m) with the number of sites in the tree and the number of sites being checked, because there are two linear searches - can get very inefficient with even a relatively small number of sites.